### ISSUE TITLE
OperationIdGenerator and graceful failure

### ISSUE THREAD

**Issue by theBradfo** [apollographql/apollo-android#4105]
**Question.**
Hello again!

We have implemented a custom `OperationIdGenerator` that reaches out to our server to register an operation.  We inconsistently do see failures in this process, whether network or server related, and would ideally like to prevent _local_ builds from failing if possible, so we've wrapped this around a conditional try/catch to make local builds not fail and assign a temporary/invalid operationId in these cases.  While this helps prevent undesirable build failures to unblock local development, these operations are effectively broken at runtime if/when they need to be executed.  

With that in mind, would it be possible to expand the functionality of the caller of `OperationIdGenerator` to be knowledgable about ignored or handled failures like this?  When assigning a temporary operationId, we'd ideally like for the _next_ build of the affected module(s) to be rerun and not cached in order to reattempt the operation registration in the future -- and lazily address the temporary/invalid operationId.

**Comment by martinbonnin** [comment]
Hi 👋 !

First thing that comes to mind is writing a file with the success/error of the registration somewhere in the `build` folder and using it to set [`task.outputs.upToDateWhen`](https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/TaskOutputs.html#upToDateWhen-groovy.lang.Closure-) to false so that the task is re-run.

Something like (wildly untested)

```kotlin
val task = tasks.named("generateServiceApolloSources").configure {
  outputs.upToDateWhen {
    file("build/generated/apollo/registrationResult").readText() == "success" 
  }
}
```

I'm hoping the above doesn't override the default up-to-date check by Gradle but I _think_ it shouldn't?

**Comment by martinbonnin** [comment]
A possible solution here would be to extract generating the operation Ids to a separate task (they are currently part of IR generation) and find a way to make this user-configurable. 

This is no trivial task and I don't want to change everyone's workflow without a clear use case. If anyone's up to this, please reach out and we can dive into this.
